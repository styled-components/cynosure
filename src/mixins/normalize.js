// @flow

// Warning: opinionatedRules were removed in normalize.css 6.0 and will be deprecated in polished 3.0.
const opinionatedRules = {
  html: {
    fontFamily: 'sans-serif',
  },

  [`a:active,
  a:hover`]: {
    outlineWidth: '0',
  },

  [`button,
  input,
  optgroup,
  select,
  textarea`]: {
    fontFamily: 'sans-serif',
    fontSize: '100%',
    lineHeight: '1.15',
  },
}

const unopinionatedRules = {
  html: {
    lineHeight: '1.15',
    textSizeAdjust: '100%',
  },

  body: {
    margin: '0',
  },

  h1: {
    fontSize: '2em',
    margin: '0.67em 0',
  },

  hr: {
    boxSizing: 'content-box',
    height: '0',
    overflow: 'visible',
  },

  pre: {
    fontFamily: 'monospace, monospace',
    fontSize: '1em',
  },

  a: {
    'background-color': 'transparent',
  },

  'abbr[title]': [
    {
      borderBottom: 'none',
      textDecoration: 'underline',
    },
    {
      textDecoration: 'underline dotted',
    },
  ],

  [`b,
  strong`]: {
    fontWeight: 'bolder',
  },

  [`code,
  kbd,
  samp`]: {
    fontFamily: 'monospace, monospace',
    fontSize: '1em',
  },

  small: {
    fontSize: '80%',
  },

  [`sub,
  sup`]: {
    fontSize: '75%',
    lineHeight: '0',
    position: 'relative',
    verticalAlign: 'baseline',
  },

  sub: {
    bottom: '-0.25em',
  },

  sup: {
    top: '-0.5em',
  },

  img: {
    borderStyle: 'none',
  },

  [`button,
  input,
  optgroup,
  select,
  textarea`]: {
    fontFamily: 'inherit',
    fontSize: '100%',
    lineHeight: '1.15',
    margin: '0',
  },

  [`button,
  input`]: {
    overflow: 'visible',
  },

  [`button,
  select`]: {
    textTransform: 'none',
  },

  [`button,
  html [type="button"],
  [type="reset"],
  [type="submit"]`]: {
    '-webkit-appearance': 'button',
  },

  [`button::-moz-focus-inner,
  [type="button"]::-moz-focus-inner,
  [type="reset"]::-moz-focus-inner,
  [type="submit"]::-moz-focus-inner`]: {
    borderStyle: 'none',
    padding: '0',
  },

  [`button:-moz-focusring,
  [type="button"]:-moz-focusring,
  [type="reset"]:-moz-focusring,
  [type="submit"]:-moz-focusring`]: {
    outline: '1px dotted ButtonText',
  },

  fieldset: {
    padding: '0.35em 0.625em 0.75em',
  },

  legend: {
    boxSizing: 'border-box',
    color: 'inherit',
    display: 'table',
    maxWidth: '100%',
    padding: '0',
    whiteSpace: 'normal',
  },

  progress: {
    verticalAlign: 'baseline',
  },

  textarea: {
    overflow: 'auto',
  },

  [`[type="checkbox"],
  [type="radio"]`]: {
    boxSizing: 'border-box',
    padding: '0',
  },

  [`[type="number"]::-webkit-inner-spin-button,
  [type="number"]::-webkit-outer-spin-button`]: {
    height: 'auto',
  },

  '[type="search"]': {
    '-webkit-appearance': 'textfield',
    outlineOffset: '-2px',
  },

  '[type="search"]::-webkit-search-decoration': {
    '-webkit-appearance': 'none',
  },

  '::-webkit-file-upload-button': {
    '-webkit-appearance': 'button',
    font: 'inherit',
  },

  details: {
    display: 'block',
  },

  summary: {
    display: 'list-item',
  },

  template: {
    display: 'none',
  },

  '[hidden]': {
    display: 'none',
  },
}

function mergeRules(baseRules: Object, additionalRules: Object): Object {
  const mergedRules = { ...baseRules }
  Object.keys(additionalRules).forEach(key => {
    if (mergedRules[key]) {
      mergedRules[key] = {
        ...mergedRules[key],
        ...additionalRules[key],
      }
    } else {
      mergedRules[key] = {
        ...additionalRules[key],
      }
    }
  })
  return mergedRules
}

/**
 * CSS to normalize abnormalities across browsers (normalize.css v8.0.0 | MIT License | github.com/necolas/normalize.css)
 *
 * @example
 * // Styles as object usage
 * const styles = {
 *    ...normalize(),
 * }
 *
 * // styled-components usage
 * injectGlobal`${normalize()}`
 *
 * // CSS as JS Output
 *
 * html {
 *   lineHeight: 1.15,
 *   textSizeAdjust: 100%,
 * } ...
 */
function normalize(excludeOpinionated?: boolean): Object {
  if (excludeOpinionated) return unopinionatedRules
  return mergeRules(unopinionatedRules, opinionatedRules)
}

export default normalize
