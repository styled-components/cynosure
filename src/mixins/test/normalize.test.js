// @flow
import normalize from '../normalize'

describe('normalize', () => {
  it('should return rules', () => {
    expect(normalize()).toMatchSnapshot()
  })
})
