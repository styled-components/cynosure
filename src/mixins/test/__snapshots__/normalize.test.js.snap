exports[`normalize should default to all rules 1`] = `
Object {
  "::-webkit-file-upload-button": Object {
    "-webkit-appearance": "button",
    "font": "inherit",
  },
  "[hidden]": Object {
    "display": "none",
  },
  "[type=\"checkbox\"],
  [type=\"radio\"]": Object {
    "boxSizing": "border-box",
    "padding": "0",
  },
  "[type=\"number\"]::-webkit-inner-spin-button,
  [type=\"number\"]::-webkit-outer-spin-button": Object {
    "height": "auto",
  },
  "[type=\"search\"]": Object {
    "-webkit-appearance": "textfield",
    "outlineOffset": "-2px",
  },
  "[type=\"search\"]::-webkit-search-cancel-button,
  [type=\"search\"]::-webkit-search-decoration": Object {
    "-webkit-appearance": "none",
  },
  "a": Object {
    "-webkit-text-decoration-skip": "objects",
    "background-color": "transparent",
  },
  "a:active,
  a:hover": Object {
    "outlineWidth": "0",
  },
  "abbr[title]": Object {
    "borderBottom": "none",
    "textDecoration": "underline dotted",
  },
  "article,
  aside,
  footer,
  header,
  nav,
  section": Object {
    "display": "block",
  },
  "audio,
  video": Object {
    "display": "inline-block",
  },
  "audio:not([controls])": Object {
    "display": "none",
    "height": "0",
  },
  "b,
  strong": Object {
    "fontWeight": "inherit",
  },
  "body": Object {
    "margin": "0",
  },
  "button,
  html [type=\"button\"],
  [type=\"reset\"],
  [type=\"submit\"]": Object {
    "-webkit-appearance": "button",
  },
  "button,
  input": Object {
    "overflow": "visible",
  },
  "button,
  input,
  optgroup,
  select,
  textarea": Object {
    "fontFamily": "sans-serif",
    "fontSize": "100%",
    "lineHeight": "1.15",
    "margin": "0",
  },
  "button,
  select": Object {
    "textTransform": "none",
  },
  "button:-moz-focusring,
  [type=\"button\"]:-moz-focusring,
  [type=\"reset\"]:-moz-focusring,
  [type=\"submit\"]:-moz-focusring": Object {
    "outline": "1px dotted ButtonText",
  },
  "button::-moz-focus-inner,
  [type=\"button\"]::-moz-focus-inner,
  [type=\"reset\"]::-moz-focus-inner,
  [type=\"submit\"]::-moz-focus-inner": Object {
    "borderStyle": "none",
    "padding": "0",
  },
  "canvas": Object {
    "display": "inline-block",
  },
  "code,
  kbd,
  samp": Object {
    "fontFamily": "monospace, monospace",
    "fontSize": "1em",
  },
  "details,
  menu": Object {
    "display": "block",
  },
  "dfn": Object {
    "fontStyle": "italic",
  },
  "fieldset": Object {
    "border": "1px solid #c0c0c0",
    "margin": "0 2px",
    "padding": "0.35em 0.625em 0.75em",
  },
  "figcaption,
  figure,
  main": Object {
    "display": "block",
  },
  "figure": Object {
    "margin": "1em 40px",
  },
  "h1": Object {
    "fontSize": "2em",
    "margin": "0.67em 0",
  },
  "hr": Object {
    "boxSizing": "content-box",
    "height": "0",
    "overflow": "visible",
  },
  "html": Object {
    "fontFamily": "sans-serif",
    "lineHeight": "1.15",
    "textSizeAdjust": "100%",
  },
  "img": Object {
    "borderStyle": "none",
  },
  "legend": Object {
    "boxSizing": "border-box",
    "color": "inherit",
    "display": "table",
    "maxWidth": "100%",
    "padding": "0",
    "whiteSpace": "normal",
  },
  "mark": Object {
    "backgroundColor": "#ff0",
    "color": "#000",
  },
  "pre": Object {
    "fontFamily": "monospace, monospace",
    "fontSize": "1em",
  },
  "progress": Object {
    "display": "inline-block",
    "verticalAlign": "baseline",
  },
  "small": Object {
    "fontSize": "80%",
  },
  "sub": Object {
    "bottom": "-0.25em",
  },
  "sub,
  sup": Object {
    "fontSize": "75%",
    "lineHeight": "0",
    "position": "relative",
    "verticalAlign": "baseline",
  },
  "summary": Object {
    "display": "list-item",
  },
  "sup": Object {
    "top": "-0.5em",
  },
  "svg:not(:root)": Object {
    "overflow": "hidden",
  },
  "template": Object {
    "display": "none",
  },
  "textarea": Object {
    "overflow": "auto",
  },
}
`;

exports[`normalize should only return unopinionated rules, when passed true 1`] = `
Object {
  "::-webkit-file-upload-button": Object {
    "-webkit-appearance": "button",
    "font": "inherit",
  },
  "[hidden]": Object {
    "display": "none",
  },
  "[type=\"checkbox\"],
  [type=\"radio\"]": Object {
    "boxSizing": "border-box",
    "padding": "0",
  },
  "[type=\"number\"]::-webkit-inner-spin-button,
  [type=\"number\"]::-webkit-outer-spin-button": Object {
    "height": "auto",
  },
  "[type=\"search\"]": Object {
    "-webkit-appearance": "textfield",
    "outlineOffset": "-2px",
  },
  "[type=\"search\"]::-webkit-search-cancel-button,
  [type=\"search\"]::-webkit-search-decoration": Object {
    "-webkit-appearance": "none",
  },
  "a": Object {
    "-webkit-text-decoration-skip": "objects",
    "background-color": "transparent",
  },
  "abbr[title]": Object {
    "borderBottom": "none",
    "textDecoration": "underline dotted",
  },
  "article,
  aside,
  footer,
  header,
  nav,
  section": Object {
    "display": "block",
  },
  "audio,
  video": Object {
    "display": "inline-block",
  },
  "audio:not([controls])": Object {
    "display": "none",
    "height": "0",
  },
  "b,
  strong": Object {
    "fontWeight": "inherit",
  },
  "button,
  html [type=\"button\"],
  [type=\"reset\"],
  [type=\"submit\"]": Object {
    "-webkit-appearance": "button",
  },
  "button,
  input": Object {
    "overflow": "visible",
  },
  "button,
  input,
  optgroup,
  select,
  textarea": Object {
    "margin": "0",
  },
  "button,
  select": Object {
    "textTransform": "none",
  },
  "button:-moz-focusring,
  [type=\"button\"]:-moz-focusring,
  [type=\"reset\"]:-moz-focusring,
  [type=\"submit\"]:-moz-focusring": Object {
    "outline": "1px dotted ButtonText",
  },
  "button::-moz-focus-inner,
  [type=\"button\"]::-moz-focus-inner,
  [type=\"reset\"]::-moz-focus-inner,
  [type=\"submit\"]::-moz-focus-inner": Object {
    "borderStyle": "none",
    "padding": "0",
  },
  "canvas": Object {
    "display": "inline-block",
  },
  "code,
  kbd,
  samp": Object {
    "fontFamily": "monospace, monospace",
    "fontSize": "1em",
  },
  "details,
  menu": Object {
    "display": "block",
  },
  "dfn": Object {
    "fontStyle": "italic",
  },
  "fieldset": Object {
    "border": "1px solid #c0c0c0",
    "margin": "0 2px",
    "padding": "0.35em 0.625em 0.75em",
  },
  "figcaption,
  figure,
  main": Object {
    "display": "block",
  },
  "figure": Object {
    "margin": "1em 40px",
  },
  "h1": Object {
    "fontSize": "2em",
    "margin": "0.67em 0",
  },
  "hr": Object {
    "boxSizing": "content-box",
    "height": "0",
    "overflow": "visible",
  },
  "html": Object {
    "lineHeight": "1.15",
    "textSizeAdjust": "100%",
  },
  "img": Object {
    "borderStyle": "none",
  },
  "legend": Object {
    "boxSizing": "border-box",
    "color": "inherit",
    "display": "table",
    "maxWidth": "100%",
    "padding": "0",
    "whiteSpace": "normal",
  },
  "mark": Object {
    "backgroundColor": "#ff0",
    "color": "#000",
  },
  "pre": Object {
    "fontFamily": "monospace, monospace",
    "fontSize": "1em",
  },
  "progress": Object {
    "display": "inline-block",
    "verticalAlign": "baseline",
  },
  "small": Object {
    "fontSize": "80%",
  },
  "sub": Object {
    "bottom": "-0.25em",
  },
  "sub,
  sup": Object {
    "fontSize": "75%",
    "lineHeight": "0",
    "position": "relative",
    "verticalAlign": "baseline",
  },
  "summary": Object {
    "display": "list-item",
  },
  "sup": Object {
    "top": "-0.5em",
  },
  "svg:not(:root)": Object {
    "overflow": "hidden",
  },
  "template": Object {
    "display": "none",
  },
  "textarea": Object {
    "overflow": "auto",
  },
}
`;
